src/spesh/disp.c

```diff
@@ -1057,6 +1110,217 @@ static int translate_dispatch_program(MVMThreadContext *tc, MVMSpeshGraph *g,
+                            tc, g, bb->linear_next, post_call_instructions->next);
+
+                        emit_iffy_op(tc, g, bb->linear_next, &post_call_instructions, MVM_OP_unless_i,
+                            rb_ins->operands[0],
+                            target_bb);
+
+                        MVMSpeshOperand box_target_reg = MVM_spesh_manipulate_new_version(tc, g, ins->operands[0].reg.orig);
+                        emit_tri_op(tc, g, bb->linear_next, &post_call_instructions, MVM_OP_box_i,
+                            box_target_reg,
+                            rb_ins->operands[0],
+                            rb_ins->operands[3]);
+
+                        MVMSpeshBB *box_bb = MVM_spesh_manipulate_split_BB_at(
+                            tc, g, bb->linear_next, post_call_instructions);
+
+                        bb->linear_next->children[1] = target_bb;
+                        bb->linear_next->succ[1] = target_bb;
+                        box_bb->num_succ = 1;
+                        box_bb->num_children = 0;
+                        box_bb->pred[0] = bb->linear_next;
+                        target_bb->num_pred = 2;
+                        target_bb->pred = MVM_spesh_alloc(tc, g, sizeof(MVMSpeshBB*) * 2);
+                        target_bb->pred[0] = bb->linear_next;
+                        target_bb->pred[1] = box_bb;
+
+                        MVM_spesh_graph_place_phi(tc, g, target_bb, 2, ins->operands[0].reg.orig);
+                        MVMSpeshIns *phi = target_bb->first_ins;
+                        phi->operands[0].reg.i = ins->operands[0].reg.i;
+                        phi->operands[1].reg.i = set_target_reg.reg.i;
+                        phi->operands[2].reg.i = box_target_reg.reg.i;
+                        MVM_spesh_get_facts(tc, g, phi->operands[0])->writer = phi;
+                        MVM_spesh_usages_add_by_reg(tc, g, set_target_reg, phi);
+                        MVM_spesh_usages_add_by_reg(tc, g, box_target_reg, phi);
+                    }
+                    else {
+                        emit_tri_op(tc, g, bb->linear_next, &post_call_instructions, MVM_OP_box_i,
+                            ins->operands[0],
+                            rb_ins->operands[0],
+                            rb_ins->operands[3]);
+                    }
+                    if (return_type_facts->flags & MVM_SPESH_FACT_KNOWN_TYPE) {
+                        MVMSpeshFacts *result_facts = MVM_spesh_get_facts(tc, g, ins->operands[0]);
+                        result_facts->flags |= MVM_SPESH_FACT_KNOWN_TYPE;
+                        result_facts->type = return_type_facts->type;
+                    }
+                    MVM_spesh_manipulate_release_temp_reg(tc, g, rb_ins->operands[0]);
+                }
+
+                /* Make sure we delay release of temporaries since optimization
+                 * can add further ones. */
+                delay_temps_release = 1;
```

Comment:
I believe this needs `optimize.c` to cooperate (to actually release them), but don't see a change there.